internal/onboarding/azure.go

```diff
@@ -245,41 +206,151 @@ func GenerateAzureHCL(integration AzureIntegration) (string, error) {
 		}
-		hclBlocks = tfgen.CombineHclBlocks(hclBlocks, vmScanningBlocks)
+		resourceRMCustomRoleDefinition := tfgen.NewResource("azurerm_role_definition", "mondoo_security",
+			tfgen.HclResourceWithAttributes(tfgen.Attributes{
+				"name":              "tf-mondoo-security-role",
+				"description":       "Allow Mondoo Security to use run commands for Virtual Machine scanning",
+				"scope":             fmt.Sprintf("/subscriptions/%s", integration.Primary),
+				"assignable_scopes": dataRMAllSubscriptions.TraverseRef("subscriptions[*]", "id"),
+			}),
+			tfgen.HclResourceWithGenericBlocks(customRolePermissionsBlock),
+		)
+		resources = append(resources, resourceRMCustomRoleDefinition)
+		dependsOn = append(dependsOn, resourceRMCustomRoleDefinition.TraverseRef())
+
+		// adds the custom role to all subscriptions
+		resourceRMCustomRoleAssignment := tfgen.NewResource("azurerm_role_assignment", "mondoo_security",
+			tfgen.HclResourceWithAttributes(tfgen.Attributes{
+				"role_definition_id": resourceRMCustomRoleDefinition.TraverseRef("role_definition_resource_id"),
+
+				"count":        tfgen.NewFuncCall("length", dataRMAllSubscriptions.TraverseRef("subscriptions")),
+				"scope":        dataRMAllSubscriptions.TraverseRef("subscriptions[count.index]", "id"),
+				"principal_id": resourceADServicePrincipal.TraverseRef("object_id"),
+			}),
+		)
+		resources = append(resources, resourceRMCustomRoleAssignment)
+		dependsOn = append(dependsOn, resourceRMCustomRoleAssignment.TraverseRef())
 	}
 
-	return tfgen.CreateHclStringOutput(hclBlocks...), nil
+	return resources, dependsOn, nil
 }
 
 // azureRMReaderRoleAssignmentBlocks creates role assignment blocks for a list of subscription IDs,
 // it returns the resources and the list of depends_on objects
-func azureRMReaderRoleAssignmentBlocks(subscriptionIDs []string, resourceADServicePrincipal *tfgen.HclResource) ([]tfgen.Object, []any) {
+func azureRMReaderRoleAssignmentBlocks(integration AzureIntegration, resourceADServicePrincipal *tfgen.HclResource) ([]tfgen.Object, []any, error) {
 	resources := []tfgen.Object{}
 	dependsOn := []any{}
 
+	// For scope format see:
+	// https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/role_assignment
+	subscriptionIDs := make([]string, len(integration.Allow))
+	for i, id := range integration.Allow {
+		subscriptionIDs[i] = fmt.Sprintf("/subscriptions/%s", id)
+	}
+
+	// custom role needed only when scanning VMs
+	var resourceRMCustomRoleDefinition *tfgen.HclResource
+	if integration.ScanVMs {
+		customRolePermissionsBlock, err := tfgen.HclCreateGenericBlock("permissions", nil,
+			tfgen.Attributes{
+				"actions ": []string{
+					"Microsoft.Compute/virtualMachines/runCommands/read",
+					"Microsoft.Compute/virtualMachines/runCommands/write",
+					"Microsoft.Compute/virtualMachines/runCommands/delete",
+					"Microsoft.Compute/virtualMachines/runCommand/action",
+				},
```

Comment:
We have this block of permissions twice now. Here and at L239.

I understand why you moved it, but perhaps keep at least the list of permissions in one place.